keps/sig-storage/1412-immutable-secrets-and-configmaps/README.md

```diff
@@ -197,6 +181,139 @@ immutable Secrets and/or ConfigMaps. That said, this is purely a performance
 improvement and doesn't have correctness implications. So those clusters will
 simple have worse scalability characteristic.
 
+## Production Readiness Review Questionnaire
+
+### Feature enablement and rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: ImmutableEphemeralVolumes
+    - Components depending on the feature gate: kube-apiserver, kubelet
+
+* **Does enabling the feature change any default behavior?**
+  No, users need to explicitly opt-in for secrets/configmaps immutability.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we rollback
+  the enablement)?**
+  Yes. kube-apiserver rollback allows for modifying secrets/configmaps.
+  kubelet rollback causes that all secrets/configmaps are again being watched.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  If the `immutable` field of Secret/ConfigMap wasn't cleared up for some
+  objects, this would result in making the immutable again.
+
+* **Are there any tests for feature enablement/disablement?**
+  Only basic conversions tests.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  If some critical Secrets/Configmaps were marked as immutable it may not be
+  able to update them if needed. No impact for already running workloads.
+
+* **What specific metrics should inform a rollback?**
+  No known rollback criteria.
+
+* **Were upgrade and rollback tested? Was upgrade->downgrade->upgrade path tested?**
+  Manual tests run to confirm that (on top of existing e2e tests):
+  - setting immutable field is impossible when the feature is switched of
+  - upgrade doesn't change behavior of existing objects
+  - immutable fields (if previously set) are not cleared after rollback (which
+    is a generic pattern used for all features)
+  - given validation is not feature-gated, it's impossible to update data even
+    after rollback; TODO: Is that a desired behavior?
```

Comment:
@thockin @liggitt - can you confirm that this is a bug that should be fixed (and cherrypicked) - i.e. that the validation should also be feature-gated?

@msau42 - who signaled this problem

Reply:
That's what I would expect. An immutable configmap was created with the feature enabled, and then the feature was disabled (either by rolling back the server or by turning off the feature gate), but the previously persisted data is honored

Reply:
I believe the current behavior of not feature gating validation means that validation will pass if the new field was previously set and we do an update/delete, which is what we want.

Reply:
@msau42 - not really; not gating validation means that: if the field was set previously, validation will forbid you to both change it as well as change the data
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/secret/strategy.go#L88
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L5069

Reply:
I think that's fine. It was set to immutable and even when rolling back, it's still immutable. We do want to allow users to delete their objects that were using the new feature.

Reply:
Deletes are allowed (as well as changing ObjectMeta, like finalizers to not break object lifecycle).

Reply:
Sounds good, I think we can remove the TODO. This is the desired behavior.

Reply:
done